API/CopilotService/__init__.py

```diff
@@ -56,15 +57,24 @@ def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
 
         if os.environ.get('ENABLE_RETRIEVAL_AUGMENTED_GENERATION', "true").lower() == "true":
             task_list, usage_context = asyncio.run(_retrieve_context_from_learn_knowledge_index(context, question))
-            question = _add_context_to_queston(context, question, task_list, usage_context)
+            token_limit = int(os.environ.get("CONTEXT_TOKEN_LIMIT", 4096))
+            completion_tokens = int(os.environ.get('OPENAI_MAX_TOKENS', 4000))   # The default value should be the same as the one in initialize_chatgpt_service_params
+            factor = float(os.environ.get('ESTIMATION_ADJUSTMENT_FACTOR', 0.95))
+            system_msg_tokens = num_tokens_from_message(system_msg) or 0
+            token_remains = (token_limit - completion_tokens) * factor - system_msg_tokens
+            question = _add_context_to_queston(context, question, task_list, usage_context, token_limit=token_remains)
 
         if service_type == ServiceType.GPT_GENERATION:
             context.custom_context.gpt_task_name = 'GENERATE_SCENARIO'
             gpt_result = gpt_generate(context, system_msg, question, history)
             result = [_build_scenario_response(gpt_result)] if gpt_result else []
 
         elif service_type == ServiceType.MIX:
-            result = knowledge_search(question, top_num)
+            try:
+                result = knowledge_search(question, top_num)
+            except HttpResponseError as e:
+                logger.error('Error from knowledge search: \n%s', e)
+                result = []
```

Comment:
I believe the same logic is also needed when `service_type == ServiceType.KNOWLEDGE_SEARCH`,

Reply:
That makes sense. Maybe we should return a status code `500` when encounting Error in this case.

Reply:
Agree